fabtests/multinode/src/timing.c

```diff
@@ -0,0 +1,191 @@
+#include "timing.h"
+#include "core.h"
+
+void multi_timer_init(struct multi_timer *timer, int rank)
+{
+	timer->rank = rank;
+	timer->start = 0;
+	timer->end = 0;
+}
+
+void multi_timer_start(struct multi_timer *timer)
+{
+	if (timer && timer->start == 0)
+		timer->start = ft_gettime_ns();
+}
+
+void multi_timer_stop(struct multi_timer *timer)
+{
+	if (timer && timer->start != 0 && timer->end == 0)
+		timer->end = ft_gettime_ns();
+}
+
+static inline void print_timer(struct multi_timer timer, char* info)
+{
+	printf("rank: %i, start: %ld, end: %ld, %s\n", 
+			timer.rank, timer.start, timer.end, info);
+}
+
+int multi_timer_analyze(struct multi_timer *timers, int timer_count)
+{
+	int i, j, ret = 0;
+	int iterations = timer_count / pm_job.num_ranks;
+	double total_timers, total_min, total_max, total_sum_time, total_duration;
+	long *min = calloc(iterations, sizeof(*min));
+	long *max = calloc(iterations, sizeof(*max));
+	long *first_start = calloc(iterations, sizeof(*first_start));
+	long *last_end = calloc(iterations, sizeof(*last_end));
+	double *sum_time = calloc(iterations, sizeof(*sum_time));
+	long duration;
+	struct multi_timer *gather_timers;
+	double iter_timer_count = 0;
+
+	gather_timers = calloc(pm_job.num_ranks * pm_job.num_ranks, 
+							sizeof(*gather_timers));
+	
+	if (!(min && max && first_start && last_end && sum_time && gather_timers))
+		goto err;
+
+	printf("%-10s %16s %16s %16s %16s\n",
+			"Iteration", "Min Send (ns)", "Max Send (ns)", 
+			"Pattern Time(ns)", "Average Send(ns)");
+
+	for (i = 0; i < iterations; i++) {
+		ret = multi_timer_iter_gather(gather_timers, timers, i);
+		if (ret < 0)
+			goto err;
+		if (pm_job.my_rank == 0) {
+			for (j = 0; j < pm_job.num_ranks * pm_job.num_ranks; j++) {
+				if (gather_timers[j].start == 0 || gather_timers[j].end == 0)
+					continue;
+
+				iter_timer_count++;
+				duration = gather_timers[j].end - gather_timers[j].start;
+				sum_time[i] += duration;
+
+				if (gather_timers[j].start < first_start[i] 
+					|| first_start[i] == 0)
+					first_start[i] = gather_timers[j].start;
+				if (gather_timers[j].end > last_end[i])
+					last_end[i] = gather_timers[j].end;
+				if (duration > max[i])
+					max[i] = duration;
+				if (duration < min[i] || min [i] == 0)
+					min[i] = duration;
+
+			}
+			printf("%-10i %16ld %16ld %16ld %16.3f\n",
+					i, min[i], max[i], last_end[i] - first_start[i],
+					sum_time[i]/iter_timer_count);
+
+			total_min += min[i];
+			total_max += max[i];
+			total_duration += last_end[i] - first_start[i];
+			total_sum_time += sum_time[i];
+			total_timers += iter_timer_count;
+		}
+		pm_barrier();
+	}
+
+	if (pm_job.my_rank == 0)
+		printf("%-10s %16.3lf %16.3lf %16.3lf %16.3lf\n", "Average",
+			total_min/iterations, total_max/iterations,
+			total_duration/iterations,
+			total_sum_time/total_timers);
+
+err:
+	free(min);
+	free(max);
+	free(sum_time);
+	free(first_start);
+	free(last_end);
+	free(gather_timers);
+	return ret & 1;
```

Comment:
This is a really weird return code.  Just return 0 on success or -error on failure.  I have no idea what the intent is here.

Reply:
So I do think I have an error here, but the reason I did this was because this ret is basically set by a sockets call and that value gets passed back to ft_exit_code, which does not like a positive value and incorrectly concludes that the tests failed. I was trying to mitigate that, but this is not the right way